[1.9] Create default_automation_condition_sensor in Definitions object

OwenKephart opened the pull request:
## Summary & Motivation

This shifts us to creating the default automation condition sensor in user-code land, which has the benefit of allowing us to update default values in sync with user-code updates, rather than host process version updates (particularly in a cloud context, these may be out of sync with each other).

However, in order to retain compatibility for users on a pre-1.9 version, we need to also keep the external data version as a fallback, in case their repository definition hasn't been generated using the latest code.

To support both use cases, I factored out the logic into a helper function.

## How I Tested These Changes

Existing tests exercise these codepaths pretty heavily in a variety of oddball situations.

## Changelog

NOCHANGELOG

OwenKephart commented:
* **#25201** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/25201?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈
* `master`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @OwenKephart and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="11px" height="11px"/> <b>Graphite</b></a>
  <!-- Current dependencies on/for this PR: -->

OwenKephart commented on python_modules/dagster/dagster/_core/definitions/utils.py:
```diff
@@ -311,3 +320,88 @@ def resolve_automation_condition(
+    return sensors
+
+
+def get_default_automation_condition_sensor_selection(
```
This ports over the logic from the old code in external.py, but I took the opportunity to simplify it a bit.

schrockn replied:
Would recommend doing simplification in a follow up PR to reduce risk

gibsondan reviewed (dismissed):
Plan seems reasonable to me 

Agree there is some useful test coverage here but i think you should consider having something that would break if you had accidentally left out the " if necessary, create a default automation condition sensor" logic in ExternalRepository? Since all definitiions are now supplying the default sensor I'm not sure we are actually exercising that code path anywhere now. For example, you could adjust test_default_automation_policy_sensors to create a frankenstein "legacy" Definitions object without that sensor (not sure how feasible that is in pratice) and verify that the ExternalRepository still supplies one?

How weird is the resulting behavior here if the Definitions is supplying a default automation condition sesor here but the deployment/instance setting to use sensors is off? I remember there was a PR that hoped to handle that case more explicitly but it would have caught it at ExternalRepository time which is a bit late in the game

OwenKephart commented:
@gibsondan re: test coverage, just added a test that constructs a repo without the auto-created sensor, and confirmed that it passes (also manually confirmed that it would fail if I removed the code in `external.py`)

> How weird is the resulting behavior here if the Definitions is supplying a default automation condition sesor here but the deployment/instance setting to use sensors is off?

This is a good callout, I think it's definitely weird but ultimately it's just more advertising for the new sensor-based world, which I think nets out as a positive (i.e. you see these sensor in your code location now).

The main concern is that now it's technically possible for you to turn this sensor on in addition to your global daemon process, which would result in duplicate runs if this was running in user code mode (if the sensors were in daemon mode, they could be turned on but they wouldn't get evaluated). To prevent this, I tossed in an invariant check that will fail the sensor tick in this scenario. Not the most elegant but it at least prevents duplicate runs.

I think I can look into getting some sort of banner in the UI if we detect that your instance setting has the old value to really make things clear.

gibsondan commented on python_modules/dagster/dagster/_core/definitions/automation_condition_sensor_definition.py:
```diff
@@ -57,6 +58,11 @@ def _evaluate(sensor_def: "AutomationConditionSensorDefinition", context: Sensor
             "or split this sensor into multiple AutomationConditionSensorDefinitions with AssetSelections that target fewer "
             "assets or checks."
         )
+    if not context.instance.auto_materialize_use_sensors:
+        raise DagsterInvalidInvocationError(
+            "Cannot evaluate an AutomationConditionSensorDefinition if the instance setting "
+            " `auto_materialize: use_sensors` is set to False. Update your configuration to prevent this error.",
+        )
```
I don't think this is going to work the way we intend here when executed in cloud user code. The source of truth for this in cloud is a deployment setting, which works fine when it is being checked in a daemon, but is not currently propagated to cloud user code. Right now the current implementation would look for the same dagster.yaml key that you would set in dagster OSS.

OwenKephart replied:
good callout, I've moved this logic to be inside the sensor daemon and added a test for the behavior

gibsondan commented on python_modules/dagster/dagster/_core/remote_representation/external.py:
```diff
@@ -466,6 +413,18 @@ def _sensor_mappings(
                 try:
                     keys = sensor.asset_selection.resolve(self.asset_graph)
                     for key in keys:
+                        # only count an asset as targeted by an automation condition sensor if it
+                        # has an automation condition
+                        if sensor.sensor_type in (
+                            SensorType.AUTO_MATERIALIZE,
+                            SensorType.AUTOMATION,
+                        ):
+                            node_snap = self.get_asset_node_snap(key)
+                            if not node_snap or not (
+                                node_snap.automation_condition
+                                or node_snap.automation_condition_snapshot
+                            ):
+                                continue
```
i'm a little unclear how this part relates to the PR - does the fact that these are now sometimes defined in user code now cause issues here?

OwenKephart replied:
I'll be honest, I don't know why i lumped this change in (it's been awhile), but the basic idea is that we use the "targeted by" to render stuff in the UI and it feels a little silly if we link to the automation condition sensor from the asset details page of an asset that doesn't have a condition on it.

gibsondan reviewed (changes requested):
everything looks good to me here except for the part where we error in user code evaluation if the setting isn't set

schrockn commented on python_modules/dagster-graphql/dagster_graphql_tests/graphql/repo.py:
```diff
@@ -2166,6 +2173,12 @@ def define_asset_checks():
 
 
 def _targets_asset_job(instigator: Union[ScheduleDefinition, SensorDefinition]) -> bool:
+    if isinstance(instigator, SensorDefinition) and instigator.sensor_type in (
+        # these contain asset selections, which are invalid with the dict repo
+        SensorType.AUTOMATION,
+        SensorType.AUTO_MATERIALIZE,
+    ):
+        return True
```
I'm not folllowing this comment and logic

OwenKephart replied:
I'll update the comment, the basic idea is that the method of constructing a repository with a dictionary is very old and does not support "unresolved asset jobs".

By that same token, it cannot support AutomationConditionSensorDefinitions (or any type of sensor/schedule that targets an asset selection).

schrockn commented on python_modules/dagster/dagster/_core/definitions/definitions_class.py:
```diff
@@ -268,10 +271,17 @@ def _create_repository_using_definitions_args(
+    # add in a default automation condition sensor definition if required
+    sensors = add_default_automation_condition_sensor(
+        sensors,
+        [asset for asset in assets if not isinstance(asset, CacheableAssetsDefinition)],
```
does this mean we don't support `CacheableAssetsDefinition` here?

OwenKephart replied:
Specifically, it means that if you explicitly construct a set of AutomationConditionSensorDefinitions with asset_selections that do not cover a CacheableAssetsDefinition with an automation condition, then that asset would not get automatically targeted by the default_automation_condition_sensor. 

So in order for this to hit you, you'd need to:
- Be using cacheable assets with automation conditions
- Have multiple AutomationConditionSensorDefinitions in your repo
- Neglect to include your cacheable assets definitions in one of those sensors

I consider this enough of an edge case (and one with an easy enough workaround -- just make sure those cacheable assets are in an asset selection), that it's not a big concern

schrockn replied:
Btw @smackesey is about to land https://github.com/dagster-io/dagster/pull/24632. Could this simplify the situation?

OwenKephart replied:
ooh nice, yes I think I could make use of that

schrockn replied:
Ok will let you sequence this but the less we have to think about `CacheableAssetsDefinition` the better

schrockn commented on python_modules/dagster/dagster/_core/definitions/utils.py:
```diff
@@ -311,3 +320,89 @@ def resolve_automation_condition(
     return list({id(obj): obj for obj in objects}.values()) if objects is not None else []
+
+
+def add_default_automation_condition_sensor(
+    sensors: Sequence["SensorDefinition"],
+    assets: Iterable[Union["AssetsDefinition", "SourceAsset"]],
+    asset_checks: Iterable["AssetsDefinition"],
+) -> Sequence["SensorDefinition"]:
+    """Adds a default automation condition sensor if the provided sensors do not already handle all
+    provided assets.
+    """
+    from dagster._core.definitions.asset_graph import AssetGraph
+    from dagster._core.definitions.automation_condition_sensor_definition import (
+        DEFAULT_AUTOMATION_CONDITION_SENSOR_NAME,
+        AutomationConditionSensorDefinition,
+    )
+
+    with disable_dagster_warnings():
+        asset_graph = AssetGraph.from_assets([*assets, *asset_checks])
+        sensor_selection = get_default_automation_condition_sensor_selection(sensors, asset_graph)
+        if sensor_selection:
+            default_sensor = AutomationConditionSensorDefinition(
+                DEFAULT_AUTOMATION_CONDITION_SENSOR_NAME, asset_selection=sensor_selection
+            )
+            sensors = [*sensors, default_sensor]
+
+    return sensors
```
what about collision with schedules?

OwenKephart replied:
only sensors can impact what asset selection we need to make here (we specifically only care about AutomationConditionSensorDefinitions)

schrockn replied:
ah I was thrown by the docblock at the top of the function. could you reword it make it more clear?

OwenKephart replied:
updated!

schrockn reviewed (dismissed):
Important bit is my question about schedules

OwenKephart commented:
### Merge activity

* **Oct 22, 4:25 PM EDT**: A user started a stack merge that includes this pull request via [Graphite](https://app.graphite.dev/github/pr/dagster-io/dagster/25201).

cnogradi commented:
This change seems to break the usage of cacheable assets that are depended on by other assets since the new sensor is added using a graph that does not contain those cacheable assets and thus it fails to resolve those assets when trying to add the new sensor.  The code that seems to cause this is here:

https://github.com/dagster-io/dagster/blob/24729d953980788d7cab1117a4c8dc2644fc64f6/python_modules/dagster/dagster/_core/definitions/definitions_class.py#L280

so before I submit a ticket for this issue, I want to make sure that there is not a work around involved where I can disable this sensor from being selected for an assets that depends on a cacheable one.  Thoughts?

OwenKephart commented:
Hi @cnogradi -- a ticket explaining more details of where you're hitting this error would be good. The intention here was that as long as you don't have _multiple_ `AutomationConditionSensorDefinitions` defined in your `Definitions` object, that all assets would be selected (cacheable or not), so if that's not happening then that seems like a bug.
